### PR TITLE
Terms improvements

### PR DESCRIPTION
**Pull request by arielb1**

### CODE REVIEW

File: specification/terms.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 Some terms, defined by @ubsan and @arielb1. Please debate on them, or correct them, as you see fit :)
 
+For explicitness, some of the section contain ADT declarations. These ADT declarations use Rust syntax for familiarity, but all smart pointers have been removed for clarity.
```
Comment (strega-nil) [inline specification/terms.md]: No longer do these ADT decls use rust syntax.

File: specification/terms.md [hunk 2]
```diff
@@ -70,28 +72,32 @@ A typed expression that evaluates to an object, as opposed to a value. The LHS o
...
 A typed expression that evaluates to an value, as opposed to an object. The RHS of an assignment and the arguments of a function are examples of vexprs.
 
-If a lexpr is used where an vexpr is expected, the memory is read from the object referred to by the lexpr to create a vexpr *if* either T: Copy, or the mutability of the lexpr allows for moving.
-
-### §2.3 Temporary lexpr
-
-An implicit lexpr that wraps a vexpr when an lexpr is expected.
+NOTE: Because of evaluation order concerns, after being potentially wrapped to a
+vexpr, subexpressions are often wrapped in a `Use(Temp(vexpr))`. The
+exact rules here should be documented.
```
Comment (strega-nil) [inline specification/terms.md:79]: Is this a Rust thing, or a rustc thing? i.e., should it be documented in the specification? Do we care that it gets put into a lexpr?

File: specification/terms.md [hunk 3]
```diff
@@ -70,28 +72,32 @@ A typed expression that evaluates to an object, as opposed to a value. The LHS o
...
 
-This case only exists when, on the right hand side of a let statement, a reference operator (`&`, `&mut`) takes the address of a vexpr. I don't know the rule. ping @nikomatsakis to actually say the rule.
+### §2.3 Temporary lexpr
 
-#### §2.3.2 Other temporary lexprs
+```
+Temp:
+    Γ ⊢ vexp VExpr
+    --------------
+    Γ ⊢ Temp(vexp) LExpr
```
Comment (strega-nil) [inline specification/terms.md]: Could you take out the `Γ ⊢`? it seems over-verbose.

File: specification/terms.md [hunk 4]
```diff
@@ -106,49 +112,92 @@ fn main() {
...
+
 ### §3.1 Capabilities
 
 All objects are able to be read, but you may not write to all objects.
 
-A lexpr associated with an object, as a vexpr, will evaluate to a value (see §4) that is equivalent to the memory of the object, interpreted as the type of the lexpr.
+A `Use` vexpr, applied to a lexpr that evaluates to an object, will evaluate to a value (see §4) that is equivalent to the memory of the object, interpreted as the type of the vexpr.
 
-A write to an object will set the memory of that object equivalent to the value of the write, without the interpretation of typing.
+A write to an object will set the memory of that object equivalent to the value of the write, without the interpretation of typing. (XXX: padding on assignment).
```
Comment (strega-nil) [inline specification/terms.md]: I know this was already here, but perhaps take out "without the interpretation of typing"?

File: specification/terms.md [hunk 5]
```diff
@@ -106,49 +112,92 @@ fn main() {
...
 
 ```rust
 let x: i32 = 0i32; // a stack, read-only object of size >= 4, alignment >= 4
 // you get the point
 ```
 
-or on the heap:
+### §3.2.3 Heap Allocations
+
+Each call to `Box::new` creates a heap allocation, which is valid until it is destroyed when the box is dropped.
```
Comment (strega-nil) [inline specification/terms.md]: There are other ways to create heap objects; Vec, HashMap, etc. Perhaps we need a more basic primitive?

**Review by strega-nil — CHANGES_REQUESTED** [review]
Few changes, ping me on irc.

File: specification/terms.md (hunk 1 above)
Comment (arielb1) [reply specification/terms.md]: Fine.

File: specification/terms.md (hunk 3 above)
Comment (arielb1) [reply specification/terms.md]: That's how the RFCs are written.

File: specification/terms.md (hunk 5 above)
Comment (arielb1) [reply specification/terms.md]: You mean `heap::alloc`? I'm for leaving everything in the heap as "implementation defined" at this stage.

Comment (strega-nil) [reply specification/terms.md]: alright, sgtm for now. leave as a FIXME?

File: specification/terms.md (hunk 3 above)
Comment (strega-nil) [reply specification/terms.md]: afaicr, it's how one RFC was written.

**Review by strega-nil — APPROVED** [review]

**Pull request merged**